extensions/amp-form/0.1/amp-form.js

```diff
@@ -209,39 +221,70 @@ export class AmpForm {
-      this.xhr_.fetchJson(xhrUrl, {
-        body: isHeadOrGet ? undefined : new FormData(this.form_),
-        method: this.method_,
-        credentials: 'include',
-        requireAmpResponseSourceOrigin: true,
-      }).then(response => {
-        this.actions_.trigger(this.form_, 'submit-success', null);
-        // TODO(mkhatib, #6032): Update docs to reflect analytics events.
-        this.analyticsEvent_('amp-form-submit-success');
-        this.setState_(FormState_.SUBMIT_SUCCESS);
-        this.renderTemplate_(response || {});
-      }).catch(error => {
-        this.actions_.trigger(this.form_, 'submit-error', null);
-        this.analyticsEvent_('amp-form-submit-error');
-        this.setState_(FormState_.SUBMIT_ERROR);
-        this.renderTemplate_(error.responseJson || {});
-        rethrowAsync('Form submission failed:', error);
+      // Wait until all variables have been substituted or 100ms timeout.
+      return this.waitOnPromisesOrTimeout_(varSubPromises, 100).then(() => {
+        if (isHeadOrGet) {
+          xhrUrl = addParamsToUrl(this.xhrAction_, this.getFormAsObject_());
+        }
+        return this.xhr_.fetchJson(xhrUrl, {
+          body: isHeadOrGet ? undefined : new FormData(this.form_),
+          method: this.method_,
+          credentials: 'include',
+          requireAmpResponseSourceOrigin: true,
+        }).then(response => {
+          this.actions_.trigger(this.form_, 'submit-success', null);
+          // TODO(mkhatib, #6032): Update docs to reflect analytics events.
+          this.analyticsEvent_('amp-form-submit-success');
+          this.setState_(FormState_.SUBMIT_SUCCESS);
+          this.renderTemplate_(response || {});
+        }).catch(error => {
+          this.actions_.trigger(this.form_, 'submit-error', null);
+          this.analyticsEvent_('amp-form-submit-error');
+          this.setState_(FormState_.SUBMIT_ERROR);
+          this.renderTemplate_(error.responseJson || {});
+          rethrowAsync('Form submission failed:', error);
+        });
       });
     } else if (this.method_ == 'POST') {
+      // non-XHR POST requests are not supported.
       if (opt_event) {
         opt_event.preventDefault();
       }
       user().assert(false,
           'Only XHR based (via action-xhr attribute) submissions are support ' +
           'for POST requests. %s',
           this.form_);
+    } else if (this.method_ == 'GET') {
+      // Non-xhr GET requests replacement should happen synchronously.
+      for (let i = 0; i < varSubsFields.length; i++) {
+        const variable = varSubsFields[i].getAttribute('default-value');
+        varSubsFields[i].value = this.urlReplacement_.expandSync(variable);
+      }
     }
   }
 
   /**
+   * Returns a race promise between resolving all promises or timing out.
+   * @param {!Array<!Promise>} promises
+   * @param {number} timeout
+   * @return {!Promise}
+   * @private
+   */
+  waitOnPromisesOrTimeout_(promises, timeout) {
+    return Promise.race(
```

Comment:
That one rejects though when timeout happens, I guess I could call the same callback in both `then` and `catch` clauses, or is there a better way to do this?